src/axom/primal/operators/detail/winding_number_3d_impl.hpp

```diff
@@ -40,191 +40,598 @@ enum class SingularityAxis
-double stokes_winding_number(const Point<T, 3>& q,
-                             const BezierCurve<T, 3>& curve,
-                             const SingularityAxis ax,
-                             int npts,
-                             double quad_tol)
+double nurbs_winding_number(const Point<T, 3>& query,
+                            const NURBSPatch<T, 3>& nPatch,
+                            const Vector<T, 3>& cast_direction,
+                            const double edge_tol = 1e-8,
+                            const double ls_tol = 1e-8,
+                            const double quad_tol = 1e-8,
+                            const double EPS = 1e-8,
+                            const int depth = 0)
 {
-  // Generate the quadrature rules in parameter space
-  static mfem::IntegrationRules my_IntRules(0, mfem::Quadrature1D::GaussLegendre);
-  const mfem::IntegrationRule& quad_rule = my_IntRules.Get(mfem::Geometry::SEGMENT, 2 * npts - 1);
+  // Skip processing of degenerate surfaces
+  if(nPatch.getNumControlPoints_u() <= 1 || nPatch.getNumControlPoints_v() <= 1)
+  {
+    return 0.0;
+  }
 
-  double quadrature = 0.0;
-  for(int qi = 0; qi < quad_rule.GetNPoints(); ++qi)
+  // Also skip processing of surfaces with zero trimming curves
+  if(nPatch.getNumTrimmingCurves() == 0)
   {
-    // Get quadrature points in space (shifted by the query)
-    const Vector<T, 3> node(q, curve.evaluate(quad_rule.IntPoint(qi).x));
-    const Vector<T, 3> node_dt(curve.dt(quad_rule.IntPoint(qi).x));
-    const double node_norm = node.norm();
+    return 0.0;
+  }
 
-    // Compute one of three vector field line integrals depending on
-    //  the orientation of the original surface, indicated through ax.
-    switch(ax)
+  const double edge_tol_sq = edge_tol * edge_tol;
+
+  // Fix the number of quadrature points arbitrarily
+  constexpr int quad_npts = 15;
+
+  // Store the winding number
+  double the_gwn = 0.0;
+
+  /* 
+   * To use Stokes theorem, we need to identify either a line containing the
+   * query that does not intersect the surface, or one that intersects the *interior*
+   * of the surface at known locations.
+   */
+
+  // Lambda to generate a 3D rotation matrix from an angle and axis
+  // Formulation from https://en.wikipedia.org/wiki/Rotation_matrix#Axis_and_angle
+  auto angleAxisRotMatrix = [](double theta, const Vector<T, 3>& axis) -> numerics::Matrix<T> {
+    const auto unitized = axis.unitVector();
+    const double x = unitized[0], y = unitized[1], z = unitized[2];
+    const double c = cos(theta), s = sin(theta), C = 1 - c;
+
+    auto matx = numerics::Matrix<T>::zeros(3, 3);
+
+    matx(0, 0) = x * x * C + c;
+    matx(0, 1) = x * y * C - z * s;
+    matx(0, 2) = x * z * C + y * s;
+
+    matx(1, 0) = y * x * C + z * s;
+    matx(1, 1) = y * y * C + c;
+    matx(1, 2) = y * z * C - x * s;
+
+    matx(2, 0) = z * x * C - y * s;
+    matx(2, 1) = z * y * C + x * s;
+    matx(2, 2) = z * z * C + c;
+
+    return matx;
+  };
```

Comment:
In case it's helpful, @BradWhitlock recently added several utility functions for generating transformations: https://github.com/LLNL/axom/blob/develop/src/axom/core/numerics/transforms.hpp

Perhaps we should move this angle-axis there? It appears a few times in axom

(No need to make this change for the current PR)